picard/coverart/plugins.py

```diff
@@ -0,0 +1,89 @@
+import json
+from hashlib import md5
+from uuid import uuid4
+from PyQt4.QtCore import QMutex
+from picard import config, log
+from picard.plugin import PluginFunctions
+
+FILE_ACTION = 0
+DOWNLOAD_ACTION = 1
+TAG_ACTION = 2
+
+_coverart_downloaded_actions = PluginFunctions()
+_coverart_tag_embed_actions = PluginFunctions()
+_coverart_file_save_actions = PluginFunctions()
+
+_process_cache = set()
+_cache_mutex = QMutex(QMutex.Recursive)
+
+
+def register_coverart_downloaded_action(action):
+    _coverart_downloaded_actions.register(action.__module__, action)
+
+
+def register_coverart_tag_embed_action(action):
+    _coverart_tag_embed_actions.register(action.__module__, action)
+
+
+def register_coverart_file_save_action(action):
+    _coverart_file_save_actions.register(action.__module__, action)
+
+
+def run_coverart_downloaded_action(coverartimage):
+    _coverart_downloaded_actions.run(coverartimage)
+
+
+def run_coverart_tag_embed_action(coverartimage):
+    _coverart_tag_embed_actions.run(coverartimage)
+
+
+def run_coverart_file_save_action(coverartimage):
+    _coverart_file_save_actions.run(coverartimage)
+
+
+_actions = [register_coverart_file_save_action,
+            register_coverart_downloaded_action,
+            register_coverart_tag_embed_action]
+
+
+class CoverartPluginFunction(object):
+
+    def __init__(self, action_type):
+        self.id = uuid4()
+        self.options = list()
+        self.processor = None
+        self.extra_settings = dict()
+        _actions[action_type](self._run_processor)
+
+    def _run_processor(self, coverartimage):
+        if not is_cached(self.id, coverartimage, self.options, self.extra_settings):
+            if self.processor:
+                self.processor(coverartimage)
+                log.debug('Running processor')
+        else:
+            log.debug('Processing cached, skipping')
+
+
+def is_cached(action_id, coverartimage, options, extra_settings):
+    _cache_mutex.lock()
+    result = False
+    try:
+        m = md5()
+        if not coverartimage.original_datahash:
+            coverartimage.original_datahash = coverartimage.datahash
+        m.update(coverartimage.original_data)
```

Comment:
I was thinking this and a solution for that deepcopy everytime is that we instead have attributes file_data and tag_data for CoverArtImage which are initialized to none and when we are saving coverart images to file/tags if they are set to none we can just use the original data and if they are not we can use them. Also we can provide a wrapper to set_data which takes care of setting data accordingly to these attributes when used by plugins.